_posts/2020-10-27-scalajs-for-scala-3.md

```diff
@@ -0,0 +1,184 @@
+
+### Reusing the .sjsir files of the Scala standard library and of the JDK subset
+
+As you may already know, a Scala 3 codebase is able to depend on libraries compiled by Scala 2.
+This is even exploited down to the standard library level: Scala 3 codebases use the standard library compiled by Scala 2.13.
+This is possible because the dotc compiler can read the "pickled" signatures in Scala 2 .class files, and it generates .class files that are binary compatible with those produced by Scala 2.
+
+Recall that, if .class files are binary compatible, corresponding .sjsir files are binary compatible as well.
+Combined with the above properties, it means that Scala 3/JS codebases can reuse libraries compiled by Scala 2/JS.
+And among them, the Scala standard library and the JDK subset.
+Therefore, can also reuse the artifacts already published on Maven Central.
+
+That property extends to the JUnit implementation, to the test bridging infrastructure, and even to any 3rd party library!
+
+### Reusing the sbt plugin
+
+Without too much detail, the Scala.js sbt plugin has 3 responsibilities:
+
+* Adding the scalajs-library.jar on the classpath of the project and the Scala.js compiler plugin
+* Provide tasks to call the linker on the .sjsir files to produce .js files
+* Set up `run` and `test` to use the produced .js file and the testing infrastructure bridge
+
+The last 2 of those responsibilities are independent of the Scala version.
+They are only concerned with .sjsir files, the linker, and the testing infrastructure, all of which can be reused.
+
+As of v1.3.0, sbt-scalajs implements the first task for Scala 2 only.
+sbt-dotty v0.4.2 and later adapts it to Scala 3, so that users need not care about anything.
+For example, it removes the Scala 2 compiler plugin that sbt-scalajs adds, and instead adds the compiler flag `-scalajs`.
+When the functionality of sbt-dotty gets merged into sbt itself, sbt-scalajs will be adapted to directly handle the setup of Scala 3.
+
+Thanks to the combination of sbt-scalajs and sbt-dotty, a straightforward build works out of the box with Scala 3 and Scala.js.
+A small example would be:
+
+```scala
+// project/plugins.sbt
+addSbtPlugin("ch.epfl.lamp" % "sbt-dotty" % "0.4.3")
+addSbtPlugin("org.scala-js" % "sbt-scalajs" % "1.3.0")
+```
+
+```scala
+// build.sbt
+ThisBuild / scalaVersion := "0.27.0-RC1"
+val myScala3JSProject = project
+  .enablePlugins(ScalaJSPlugin)
+  .settings(
+    // Scala.js project with a main method
+    scalaJSUseMainModuleInitializer := true,
+  )
+```
+
+### Reusing the tests
+
+While reusing all of the above was definitely a huge time saver (several person·years), it is the reuse of tests that I valued the most during this work.
+Have you ever dreamed about implementing a significant system from scratch without having to write the tests, because the tests already exist?
+If yes, you understand what I mean!
+
```

Comment:
Love this paragraph.

I remember telling my boss on the NetLogo project that if the devil told me he was going to either delete our compiler or delete our test suite and offered me the choice, I would far rather he delete the compiler, I wouldn't even hesitate, no contest!